improvement(perf tests): send emails from pipeline

fgelcer opened the pull request:
moving the send email calls to a separated pipeline
step, rather than sending it during the test itself.

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

fgelcer commented:
@aleksbykov @dkropachev , all checks were fixed, so it is good to go

aleksbykov commented on sct.py:
```diff
@@ -932,7 +939,31 @@ def collect_logs(test_id=None, logdir=None, backend=None, config_file=None):
+            content['attachments'] = []
+        if 'template' not in content:
+            content['template'] = None
+        email_content = content.copy()
```
maybe it is better to use deepcopy()? because copy() doesn't copy nested list/dicts starting from level 2

fgelcer replied:
done

aleksbykov commented on sct.py:
```diff
@@ -963,50 +996,62 @@ def send_email(test_id=None, test_status=None, start_time=None, started_by=None,
         with open(os.path.join(testrun_dir, 'test_id'), 'r', encoding='utf-8') as file:
             test_id = file.read().strip()
         email_results_file = os.path.join(testrun_dir, "email_data.json")
+        if not os.path.exists(email_results_file):
+            email_results_file = "email_data.json" if os.path.exists("email_data.json") else None
+        if not email_results_file:
+            LOGGER.error("Results file not found")
+            sys.exit(1)
```
Why this code needed?

aleksbykov replied:
May it is better to save under the same file like for longevity?

fgelcer replied:
in case there is no `email_results_file` the script will exit with `RC=1`

fgelcer replied:
> May it is better to save under the same file like for longevity?

it is based on the same logic as longevity tests, but because of performance tests run on builders, and not SCT runners, it might differ a bit, but same logic for both cases

aleksbykov replied:
am i right, that the email_data.json is stored under scylla-cluster-tests/ dir on builder? Let store it under test log dir

fgelcer replied:
> am i right, that the email_data.json is stored under scylla-cluster-tests/ dir on builder? Let store it under test log dir

one issue i see (at least on builders) is that for `--logdir` is sent as `--logdir "`pwd`"` so it won't be in the right path anyway

aleksbykov commented on sdcm/report_templates/results_base_custom.html:
```diff
@@ -192,6 +192,19 @@ <h3>Links:</h3>
+        </ul>
+    </div>
+    {% endif %}
+    <h3>Snapshots:</h3>
```
this line looks not relevant

fgelcer replied:
you are right, fixed

fgelcer commented:
@aleksbykov , i also changed the following line:
```
if 'latency' in os.environ.get('JOB_NAME') or 'throughput' in os.environ.get('JOB_NAME'):
```

to:
```
test_config = os.environ.get('test_config')
    if 'latency' in test_config or 'throughput' in test_config or 'perf' in test_config:
```

fgelcer commented:
> @aleksbykov , i also changed the following line:
> 
> ```
> if 'latency' in os.environ.get('JOB_NAME') or 'throughput' in os.environ.get('JOB_NAME'):
> ```
> 
> to:
> 
> ```
> test_config = os.environ.get('test_config')
>     if 'latency' in test_config or 'throughput' in test_config or 'perf' in test_config:
> ```

reverted it, as inside the `SCT_RUNNER` the environment variables are not the same as inside the `BUILDER`

aleksbykov commented on sdcm/results_analyze/__init__.py:
```diff
@@ -604,8 +602,23 @@ def check_regression(self, test_id, is_gce=False, email_subject_postfix=None, us
             subject = f'{subject} {email_subject_postfix}'
-        html = self.render_to_html(results)
-        self.send_email(subject, html)
+
+        email_data = {'email_body': results,
+                      'attachments': (),
+                      'template': self._email_template_fp}
+
+        file_path = 'email_data.json'
+        if os.path.exists(file_path):
+            with open(file_path, 'r') as file:
+                data = file.read().strip()
+                file_content = json.loads(data or '{}')
+        else:
+            file_content = dict()
+
+        file_content[subject] = email_data.copy()
+
+        with open(file_path, 'w') as file:
+            json.dump(file_content, file)
```
This code repeated in each check_regression method. May it is better to create new method in Base result analyzer with nae save_results_for_email or prepare_email

fgelcer replied:
dope

aleksbykov commented on sdcm/results_analyze/__init__.py:
```diff
@@ -802,11 +815,28 @@ def check_regression_with_subtest_baseline(self, test_id, base_test_id, subtest_
+                      'attachments': (),
+                      'template': template}
+
+        file_path = 'email_data.json'
+        if os.path.exists(file_path):
+            with open(file_path, 'r') as file:
+                data = file.read().strip()
+                file_content = json.loads(data or '{}')
+        else:
+            file_content = dict()
+
+        file_content[subject] = email_data.copy()
+
+        with open(file_path, 'w') as file:
+            json.dump(file_content, file)
```
same code

fgelcer replied:
done

aleksbykov commented on sdcm/results_analyze/__init__.py:
```diff
@@ -219,25 +210,39 @@ def check_regression(self, test_id, data, is_gce=False):  # pylint: disable=too-
+                                                  template_file='results_reactor_stall_events_list.html')
+                           ]
+
+        email_data = {'email_body': results,
+                      'attachments': attachment_file,
+                      'template': self._email_template_fp}
+
+        file_path = 'email_data.json'
+        if os.path.exists(file_path):
+            with open(file_path, 'r') as file:
+                data = file.read().strip()
+                file_content = json.loads(data or '{}')
+        else:
+            file_content = dict()
+
+        file_content[subject] = email_data.copy()
+
+        with open(file_path, 'w') as file:
+            json.dump(file_content, file)
+
+        return True
```
same here

fgelcer replied:
done

aleksbykov commented on sdcm/results_analyze/__init__.py:
```diff
@@ -1165,8 +1195,25 @@ def check_regression_multi_baseline(
-        self.send_email(email_subject, html)
+            template = 'results_performance_multi_baseline.html'
+
+        email_data = {'email_body': results,
+                      'attachments': (),
+                      'template': template}
+
+        file_path = 'email_data.json'
+        if os.path.exists(file_path):
+            with open(file_path, 'r') as file:
+                data = file.read().strip()
+                file_content = json.loads(data or '{}')
+        else:
+            file_content = dict()
+
+        file_content[subject] = email_data.copy()
+
+        with open(file_path, 'w') as file:
+            json.dump(file_content, file)
+
```
same here

fgelcer replied:
done

aleksbykov commented on sdcm/send_email.py:
```diff
@@ -629,8 +629,9 @@ def read_email_data_from_file(filename):
         try:
-            with open(filename, "r") as fp:  # pylint: disable=invalid-name
-                email_data = json.load(fp)  # pylint: disable=invalid-name
+            with open(filename, "r") as file:
+                data = file.read().strip()
+                email_data = json.loads(data or '{}')
```
Why code was changed? is it need to read file content to variable and then convert string to json?

fgelcer replied:
it happened couple of times, when trying to read a file (as a test can have multiple tests in it - not the same as the parallel test) and because of the content of the file was empty (file existed), then `json` library failed to `load` a `None` object... this is a way to fix it, and if there is no data in the file, we return `{}`, that will give us an empty dictionary

aleksbykov commented on vars/perfRegressionParallelPipeline.groovy:
```diff
@@ -230,6 +230,29 @@ def call(Map pipelineParams) {
+                                                            set -xe
+                                                            env
+
+                                                            test_id=`cat latest/test_id`
+                                                            echo "start send-email stage ..."
+                                                            ./docker/env/hydra.sh send-email --test-id "`cat latest/test_id`" --email-recipients $email_recipients --logdir "`pwd`"
```
you have variable `test_id`, you can use it in command,

fgelcer replied:
where do i have `test_id` variable?
in the environment variables it doesn't exist, and in the groovy file there is no declaration of it...
if we do, please, point it out for me please

aleksbykov replied:
then this line is not needed i think `test_id=`cat latest/test_id`

fgelcer replied:
removing

fgelcer commented on sct.py:
```diff
@@ -657,17 +657,24 @@ def conf_docs(output_format):
+    if not test_results:
+        LOGGER.error("Test Results file not found")
+        sys.exit(1)
+    if not emails:
+        LOGGER.warning("No email recipients. Email will not be sent")
+        sys.exit(1)
```
ok

fgelcer commented on sct.py:
```diff
@@ -963,50 +996,63 @@ def send_email(test_id=None, test_status=None, start_time=None, started_by=None,
-            test_results.update({
-                "test_id": test_id,
-                "nodes": get_running_instances_for_email_report(test_id, runner_ip)
+        if test_results:
+            reporter = test_results.get("reporter", "")
+            test_results['nodes'] = get_running_instances_for_email_report(test_id, runner_ip)
```
right, fixed

fgelcer commented on sct.py:
```diff
@@ -932,7 +939,31 @@ def collect_logs(test_id=None, logdir=None, backend=None, config_file=None):
 
 
+# pylint: disable=too-many-locals,too-many-statements,too-many-arguments
+def send_perf_email(test_id, email_recipients, test_results, testrun_dir, start_time, es_doc_type='test_stats'):
```
it is a bit problematic to move it to `send_email` module, because it will create a circular import

fgelcer replied:
fixed

fgelcer commented:
@aleksbykov ping

aleksbykov commented on sct.py:
```diff
@@ -963,50 +973,63 @@ def send_email(test_id=None, test_status=None, start_time=None, started_by=None,
-        test_results['nodes'] = get_running_instances_for_email_report(test_id, runner_ip)
+    job_name = os.environ.get('JOB_NAME', '')
+    if 'latency' in job_name or 'throughput' in job_name or 'perf' in job_name:
+        if not test_results:
+            LOGGER.error("Test Results file not found")
+            sys.exit(1)
+        reporter = BaseResultsAnalyzer(es_index=test_id, es_doc_type='test_stats', email_recipients=email_recipients)
```
do we need send any email if no results found for performance job?

aleksbykov commented on sdcm/results_analyze/__init__.py:
```diff
@@ -146,41 +145,45 @@ def render_to_html(self, results, html_file_path="", template=None):
         return "%s/%s" % (self._conf.get('kibana_url'), dashboard_path)
 
+    @staticmethod
+    def save_email_data_file(subject, email_data, file_path='email_data.json'):
+        if os.path.exists(file_path):
+            with open(file_path, 'r') as file:
+                data = file.read().strip()
+                file_content = json.loads(data or '{}')
+        else:
+            file_content = dict()
+        file_content[subject] = email_data.copy()
+        with open(file_path, 'w') as file:
+            json.dump(file_content, file)
+
```
Need some log debug or info messages about the progress. So if something go wrong, it was easy to find what was a problem

fgelcer replied:
added

aleksbykov reviewed (dismissed):
LGTM, small fixes required

fgelcer commented:
@aleksbykov , please confirm the requested changes were addressed

fgelcer commented:
> @aleksbykov , please confirm the requested changes were addressed

the negative path has worked fine, please, re-review

aleksbykov reviewed (approved):
Latest small comment

fgelcer commented:
@roydahan @dkropachev

fgelcer commented:
@roydahan ping

fgelcer commented:
> @fgelcer please check the precommit failures.

@roydahan , fixed

fgelcer commented:
ping !!

fgelcer commented:
> @fgelcer which jobs are using branch-perf-v10?

latency with nemesis

fgelcer commented:
ping @scylladb/qa-maintainers / @roydahan

roydahan commented:
There are too many conflicts for this to get in.
We may want to bump the version of performance in all/most perf jobs anyway.

fgelcer commented:
> There are too many conflicts for this to get in. We may want to bump the version of performance in all/most perf jobs anyway.

i will try to back port, but i think it is time to move forward and rebase `branch-perf-v11` on latest master, and start setting all perf jobs to use it, as many useful features were introduced

fgelcer commented:
> > There are too many conflicts for this to get in. We may want to bump the version of performance in all/most perf jobs anyway.
> 
> i will try to back port, but i think it is time to move forward and rebase `branch-perf-v11` on latest master, and start setting all perf jobs to use it, as many useful features were introduced

@roydahan , it will take some time to be able to resolve all the conflicts (they seem to be related to newer features in master and not to the PR itself)...
LMK when you want to move forward the branch of the perf jobs
